File: bindings/csharp/WickraGym/Env.cs
using System.Runtime.InteropServices;
using System.Text;

namespace Wickra.Gym;

/// <summary>
/// A deterministic, Gymnasium-compatible backtest environment over the wickra-gym
/// C ABI. Construct from a spec JSON, drive it with command JSONs (<c>load</c>,
/// <c>reset</c>, <c>step</c>, <c>spec</c>, <c>version</c>), and dispose it.
/// </summary>
public sealed class Env : IDisposable
{
    private readonly EnvHandle _handle;

    /// <summary>Construct an environment from an <c>EnvSpec</c> JSON string.</summary>
    /// <exception cref="ArgumentException">The spec fails to parse or validate.</exception>
    public Env(string specJson)
    {
        IntPtr raw = Native.wickra_gym_new(Utf8(specJson));
        if (raw == IntPtr.Zero)
        {
            throw new ArgumentException("wickra-gym: invalid spec", nameof(specJson));
        }
        _handle = new EnvHandle(raw);
    }

    /// <summary>Apply a command JSON and return the response JSON string.</summary>
    public string Command(string cmdJson)
    {
        byte[] cmd = Utf8(cmdJson);
        bool added = false;
        try
        {
            _handle.DangerousAddRef(ref added);
            IntPtr handle = _handle.DangerousGetHandle();

            // Start with a large buffer so a single call suffices. On the rare
            // re-alloc path, the core serves the cached response of the same
            // not-yet-delivered command, so a mutating command is never re-run.
            byte[] buf = new byte[65536];
            int n = Native.wickra_gym_command(handle, cmd, buf, (nuint)buf.Length);
            if (n < 0)
            {
                throw new InvalidOperationException("wickra-gym: command error");
            }
            if (n >= buf.Length)
            {
                buf = new byte[n + 1];
                n = Native.wickra_gym_command(handle, cmd, buf, (nuint)buf.Length);
                if (n < 0)
                {
                    throw new InvalidOperationException("wickra-gym: command error");
                }
            }
            return Encoding.UTF8.GetString(buf, 0, n);
        }
        finally
        {
            if (added)
            {
                _handle.DangerousRelease();
            }
        }
    }

    /// <summary>The wickra-gym version string.</summary>
    public static string Version() =>
        Marshal.PtrToStringUTF8(Native.wickra_gym_version()) ?? string.Empty;

    /// <summary>Release the native environment handle.</summary>
    public void Dispose() => _handle.Dispose();

    private static byte[] Utf8(string s)
    {
        byte[] b = Encoding.UTF8.GetBytes(s);
        Array.Resize(ref b, b.Length + 1); // NUL-terminate for the C ABI.
        return b;
    }
}

/// <summary>A <see cref="SafeHandle"/> owning a <c>WickraGymEnv*</c>.</summary>
internal sealed class EnvHandle : SafeHandle
{
    public EnvHandle(IntPtr handle)
        : base(IntPtr.Zero, ownsHandle: true) => SetHandle(handle);

    public override bool IsInvalid => handle == IntPtr.Zero;

    protected override bool ReleaseHandle()
    {
        Native.wickra_gym_free(handle);
        return true;
    }
}

File: bindings/java/src/main/java/org/wickra/gym/Env.java
package org.wickra.gym;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;

/**
 * A deterministic, Gymnasium-compatible backtest environment over the wickra-gym
 * C ABI (FFM/Panama). Construct from a spec JSON, drive it with command JSON
 * ({@code load}, {@code reset}, {@code step}, {@code spec}, {@code version}) and
 * read back the response JSON — the same protocol as every other binding.
 */
public final class Env implements AutoCloseable {
    private MemorySegment handle;

    /**
     * Construct an environment from an {@code EnvSpec} JSON string.
     *
     * @throws IllegalArgumentException the spec fails to parse or validate
     */
    public Env(String specJson) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment spec = arena.allocateFrom(specJson);
            MemorySegment created = (MemorySegment) Native.NEW.invokeExact(spec);
            if (created.address() == 0) {
                throw new IllegalArgumentException("wickra-gym: invalid spec");
            }
            this.handle = created;
        } catch (RuntimeException | Error e) {
            throw e;
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /**
     * Apply a command JSON and return the response JSON. Starts with a large
     * buffer so a single call suffices; on the rare re-alloc path the core serves
     * the cached response of the same not-yet-delivered command, so a mutating
     * command is never re-run. Domain errors come back in-band as
     * {@code {"ok":false,...}} JSON, not as an exception.
     */
    public String command(String cmdJson) {
        if (handle == null) {
            throw new IllegalStateException("env is closed");
        }
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment cmd = arena.allocateFrom(cmdJson);
            long cap = 65536L;
            MemorySegment buf = arena.allocate(cap);
            int n = (int) Native.COMMAND.invokeExact(handle, cmd, buf, cap);
            if (n < 0) {
                throw new IllegalStateException("wickra-gym: command failed (code " + n + ")");
            }
            if (n >= cap) {
                cap = (long) n + 1;
                buf = arena.allocate(cap);
                n = (int) Native.COMMAND.invokeExact(handle, cmd, buf, cap);
                if (n < 0) {
                    throw new IllegalStateException("wickra-gym: command failed (code " + n + ")");
                }
            }
            return buf.getString(0);
        } catch (RuntimeException | Error e) {
            throw e;
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /** The wickra-gym version. */
    public static String version() {
        try {
            MemorySegment ptr = (MemorySegment) Native.VERSION.invokeExact();
            return ptr.reinterpret(Long.MAX_VALUE).getString(0);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /** Free the native environment handle. */
    @Override
    public void close() {
        if (handle != null) {
            try {
                Native.FREE.invokeExact(handle);
            } catch (Throwable t) {
                throw new RuntimeException(t);
            }
            handle = null;
        }
    }
}
